src/TraceEvent/BPerf/BPerfEventSource.cs

```diff
@@ -0,0 +1,312 @@
+﻿//     Copyright (c) Microsoft Corporation.  All rights reserved.
+// This file is best viewed using outline mode (Ctrl-M Ctrl-O)
+//
+// This program uses code hyperlinks available as part of the HyperAddin Visual Studio plug-in.
+// It is available from http://www.codeplex.com/hyperAddin 
+// 
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Microsoft.Diagnostics.Tracing.Parsers;
+using Microsoft.Diagnostics.Tracing.Parsers.Kernel;
+
+namespace Microsoft.Diagnostics.Tracing
+{
+    public sealed class BPerfEventSource : TraceEventDispatcher, IDisposable
+    {
+        private const int OffsetToExtendedData = 88; // offsetof(TraceEventNativeMethods.EVENT_RECORD*, ExtendedData);
+
+        private const ushort CompressionFormatXpressHuff = 4;
+
+        private const int BufferSize = 1024 * 1024 * 2;
+
+        private const int ReadAheadBufferSize = 1024 * 1024 * 1;
+
+        private static readonly Guid EventTraceGuid = new Guid("{68fdd900-4a3e-11d1-84f4-0000f80464e3}");
+
+        private readonly string btlFilePath;
+
+        private readonly long fileOffset;
+
+        private readonly byte[] workspace;
+
+        private readonly byte[] uncompressedBuffer;
+
+        private readonly byte[] compressedBuffer;
+
+        private readonly Dictionary<int, string> processNameForID = new Dictionary<int, string>();
+
+        private int eventsLost;
+
+        public BPerfEventSource(string btlFilePath)
+            : this(btlFilePath, 0)
+        {
+        }
+
+        public BPerfEventSource(string btlFilePath, long fileOffset)
+            : this(btlFilePath, fileOffset, new byte[BufferSize], new byte[BufferSize], new byte[BufferSize])
+        {
+        }
+
+        public BPerfEventSource(string btlFilePath, long fileOffset, byte[] workspace, byte[] uncompressedBuffer, byte[] compressedBuffer)
+        {
+            this.btlFilePath = btlFilePath;
+            this.fileOffset = fileOffset;
+            this.workspace = workspace;
+            this.uncompressedBuffer = uncompressedBuffer;
+            this.compressedBuffer = compressedBuffer;
+
+            this.ProcessInner();
+        }
+
+        public override int EventsLost => this.eventsLost;
+
+        public override long Size => File.Exists(this.btlFilePath) ? new FileInfo(this.btlFilePath).Length : 0;
+
+        public override bool Process()
+        {
+            this.ProcessInner();
+            return true;
+        }
+
+        internal override string ProcessName(int processID, long time100ns)
+        {
+            if (!this.processNameForID.TryGetValue(processID, out var ret))
```

Comment:
Does this format handle multiple processes?  If so, this won't work if PIDs are re-used.

Reply:
It does but the calls to this are made in such a way that the dictionary can only have a single PID at any given time.